src/tir/transforms/inject_software_pipeline.cc

```diff
@@ -494,6 +512,267 @@ class PipelineRewriter : public StmtExprMutator {
+      // done by the previous iteration, so its wait_count is calculated as ((i - 1) + 3) - i. The
+      // sum of the two wait_counts gives 5.
+
+      auto& dep_local_state = (*async_states_local)[producer_stage_idx];
+      const auto num_commit_group = dep_local_state.commit_groups.size();
+      std::vector<Optional<PrimExpr>> producer_head_per_commit;
+
+      if (num_commit_group == 0) {
+        // Epilogue, no async producer. Since "local" producer_head is not available, use
+        // "global" producer_head.
+        ICHECK(!dep_local_state.producer_head);
+        producer_head_per_commit.push_back(async_states[producer_stage_idx].producer_head);
+      } else {
+        ICHECK(dep_local_state.producer_head);
+        std::vector<bool> need_wait_count(num_commit_group, true);
+
+        for (auto read_region : new_blocks[i].block->reads) {
+          if (!async_states[producer_stage_idx].writes(read_region->buffer)) continue;
+          auto commit_group_id = buffer_to_commit_group.at(read_region->buffer.get());
+          if (!need_wait_count[commit_group_id]) continue;
+
+          if (!dep_local_state.seen.count(read_region->buffer.get())) {
+            // Multiple async producers interleaved: The most recent async write is from the
+            // previous iteration. This is the B_shared case above.
+            producer_head_per_commit.push_back(dep_local_state.producer_head.value() - 1);
+          } else {
+            // Normal case
+            producer_head_per_commit.push_back(dep_local_state.producer_head.value());
+          }
+
+          need_wait_count[commit_group_id] = false;
+        }
+      }
+
+      auto wait_count = [=, &ana_normalized]() {
+        auto sum = PrimExpr(0);
+        for (auto producer_head : producer_head_per_commit) {
+          if (producer_head && ana_normalized->CanProve(producer_head.value() >= 0)) {
+            // Here, new_blocks[i].access_index corresponds to "consumer_head".
+            // The difference of producer_head and consumer_head is precisely the number of
+            // async commit groups that can still be in flight after this wait.
+            sum += analyzer_.Simplify(producer_head.value() - new_blocks[i].access_index);
+          } else {
+            // The precise count cannot be determined, give up.
+            return PrimExpr(0);
+          }
+        }
+        return sum;
+      }();
+
+      auto& pending_wait = dep_local_state.pending_wait;
+
+      if (!pending_wait.valid()) {
+        pending_wait = {static_cast<int>(i), wait_count};
+      } else if (analyzer_.CanProve(wait_count < pending_wait.wait_count)) {
+        // Coalesce multiple wait_queue if the later one allows fewer in-flight ops.
+        pending_wait = {pending_wait.insert_before, wait_count};
+      }
+    }
+  }
+
+  // Given pipelined blocks and async-related information, generate final loop statements with async
+  // scopes (if any).
+  Array<Stmt> CompletePipelineLoopStatements(
```

Comment:
I'm not entirely happy with the choice of this name, a suggestion for better one welcome.